src/main/kotlin/com/exactpro/th2/conn/dirty/fix/brokenconn/strategy/StrategyState.kt

```diff
@@ -34,46 +33,79 @@ class StrategyState(val config: RuleConfiguration? = null) {
-
+    val messageIDs: MutableList<MessageID> = Collections.synchronizedList(ArrayList<MessageID>())
 
     private val writeLock = ReentrantLock()
-    private val missedMessagesCache: MutableMap<Int, ByteBuf> = ConcurrentHashMap<Int, ByteBuf>()
-
+    private val missedMessagesCache: MutableMap<Long, ByteBuf> = ConcurrentHashMap<Long, ByteBuf>()
     private val batchMessageCacheSize = AtomicInteger(0)
-    fun getBatchMessageCacheSize() = batchMessageCacheSize.get()
 
     private val missedIncomingMessagesCount = AtomicInteger(0)
-    fun getMissedIncomingMessagesCount() = missedIncomingMessagesCount.get()
-
-    private val missedOutgoingMessagesCount = AtomicInteger(0)
-    fun getMissedOutgoingMessagesCount() = missedOutgoingMessagesCount.get()
-
-    val transformedIncomingMessagesCount = AtomicInteger(0)
+    fun updateMissedIncomingMessagesCountIfCondition(condition: (Int) -> Boolean): Boolean = writeLock.withLock {
+        var updated = false
+        missedIncomingMessagesCount.updateAndGet {
+            if(condition(it + 1)) {
+                updated = true
+                it + 1
+            } else {
+                it
+            }
+        }
+        updated
+    }
 
-    fun incrementMissedIncomingMessages() { missedIncomingMessagesCount.incrementAndGet() }
-    fun incrementIncomingTransformedMessages() { transformedIncomingMessagesCount.incrementAndGet() }
+    private val transformedIncomingMessagesCount = AtomicInteger(0)
+    fun getTransformedIncomingMessagesCount() = writeLock.withLock { transformedIncomingMessagesCount.get() }
+    fun transformIfCondition(condition: (Int) -> Boolean, transform: () -> Unit): Boolean = writeLock.withLock {
+        var updated = false
+        transformedIncomingMessagesCount.updateAndGet {
+            if(condition(it + 1)) {
+                updated = true
+                transform()
+                it + 1
+            } else {
+                it
+            }
+        }
+        updated
+    }
 
-    fun addMissedMessageToCache(sequence: Int, message: ByteBuf) {
+    private val missedOutgoingMessagesCount = AtomicInteger(0)
+    fun addMissedMessageToCacheIfCondition(sequence: Long, message: ByteBuf, condition: (Int) -> Boolean): Boolean {
         writeLock.withLock {
-            missedMessagesCache[sequence] = message
-            missedOutgoingMessagesCount.incrementAndGet()
+            var updated = false
+            missedOutgoingMessagesCount.updateAndGet {
+                if(condition(it + 1)) {
+                    missedMessagesCache[sequence] = message
+                    updated = true
+                    it + 1
+                } else {
+                    it
+                }
+            }
+            return updated
         }
     }
 
-    fun getMissedMessage(sequence: Int): ByteBuf? = missedMessagesCache[sequence]
+    fun getMissedMessage(sequence: Long): ByteBuf? = missedMessagesCache[sequence]
```

Comment:
We must be consistent with locks. Either we use the lock on read and write or we don't use the lock at all. There is no point of use the lock only for writing.

There is a `ReentrantReadWriteLock` that you can use here